File: 2nd Year/2nd Semester/MPP/PROIECT-C#/Lab1/TriatlonPersistance/RezultatRepository.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SQLite;
using TriatlonModel;
using log4net;

namespace TriatlonPersistance
{
    public class RezultatRepository : IRezultatRepository
    {
        private static readonly ILog logger = LogManager.GetLogger(typeof(RezultatRepository));
        public readonly string connectionString;

        public RezultatRepository()
        {
            connectionString = ConfigurationManager.ConnectionStrings["DefaultConnection"].ConnectionString;
        }
        public void Delete(int id)
        {
            logger.Debug("Enter deleting Rezultat with id: " + id);
            string sql = "DELETE FROM Rezultat WHERE id=@id";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        command.Parameters.AddWithValue("@id", id);
                        command.ExecuteNonQuery();
                        logger.Debug("Deleted Rezultat with id: " + id);
                    }
                    connection.Close();
                }
                logger.Debug("Exit deleting Rezultat with id: " + id);
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error deleting Rezultat with id: " + id, ex);
            }
        }

        public List<Rezultat> FindAll()
        {
            logger.Debug("Enter FindAll Rezultat");
            List<Rezultat> rezultate = new List<Rezultat>();
            string sql = "SELECT * FROM Rezultat";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        using (SQLiteDataReader reader = command.ExecuteReader())
                        {
                            while (reader.Read())
                            {
                                int participant_id = reader.GetInt32(1);
                                Participant participant = new ParticipantRepository().FindOne(participant_id);
                                int arbitru_id = reader.GetInt32(2);
                                Arbitru arbitru = new ArbitruRepository().FindOne(arbitru_id);
                                String tipProbaStr = reader.GetString(3);
                                TipProba tipProba = (TipProba)Enum.Parse(typeof(TipProba), tipProbaStr);
                                int points = reader.GetInt32(4);
                                Rezultat rezultat = new Rezultat(reader.GetInt32(0), participant, arbitru, tipProba, points);
                                rezultate.Add(rezultat);
                                logger.Debug("Found Rezultat: " + rezultat);
                            }
                        }
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error finding Rezultate", ex);
            }
            logger.Debug("Exit FindAll Rezultat");
            return rezultate;
        }

        public Rezultat FindOne(int id)
        {
            logger.Debug("Enter FindOne Rezultat with id: " + id);
            string sql = "SELECT * FROM Rezultat WHERE id=@id";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        command.Parameters.AddWithValue("@id", id);
                        using (SQLiteDataReader reader = command.ExecuteReader())
                        {
                            if (reader.Read())
                            {
                                int participant_id = reader.GetInt32(1);
                                Participant participant = new ParticipantRepository().FindOne(participant_id);
                                int arbitru_id = reader.GetInt32(2);
                                Arbitru arbitru = new ArbitruRepository().FindOne(arbitru_id);
                                String tipProbaStr = reader.GetString(3);
                                TipProba tipProba = (TipProba)Enum.Parse(typeof(TipProba), tipProbaStr);
                                int points = reader.GetInt32(4);
                                Rezultat rezultat = new Rezultat(reader.GetInt32(0), participant, arbitru, tipProba, points);
                                logger.Debug("Found Rezultat: " + rezultat);
                                return rezultat;
                            }
                        }
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error finding Rezultat with id: " + id, ex);
            }
            logger.Debug("Exit FindOne Rezultat with id: " + id);
            return null;
        }

        public void Save(Rezultat entity)
        {
            logger.Debug("Enter Save Rezultat: " + entity);
            string sql = "INSERT INTO Rezultat (participant_id, arbitru_id, tip_proba, punctaj) VALUES (@participant_id, @arbitru_id, @tipProba, @points)";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        command.Parameters.AddWithValue("@participant_id", entity.Participant.Id);
                        command.Parameters.AddWithValue("@arbitru_id", entity.Arbitru.Id);
                        command.Parameters.AddWithValue("@tipProba", entity.TipProba.ToString());
                        command.Parameters.AddWithValue("@points", entity.Points);
                        int affectedRows = command.ExecuteNonQuery();
                        if (affectedRows > 0)
                        {
                            command.CommandText = "SELECT last_insert_rowid()";
                            entity.Id = Convert.ToInt32(command.ExecuteScalar());
                        }
                        logger.Debug("Saved Rezultat: " + entity);
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error saving Rezultat: " + entity, ex);
            }
            logger.Debug("Exit Save Rezultat: " + entity);
        }

        public void Update(Rezultat entity)
        {
            logger.Debug("Enter Update Rezultat: " + entity);
            string sql = "UPDATE Rezultat SET participant_id=@participant_id, arbitru_id=@arbitru_id, tip_proba=@tipProba, punctaj=@points WHERE id=@id";
            try
            {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString))
                {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection))
                    {
                        command.Parameters.AddWithValue("@participant_id", entity.Participant.Id);
                        command.Parameters.AddWithValue("@arbitru_id", entity.Arbitru.Id);
                        command.Parameters.AddWithValue("@tipProba", entity.TipProba.ToString());
                        command.Parameters.AddWithValue("@points", entity.Points);
                        command.Parameters.AddWithValue("@id", entity.Id);
                        command.ExecuteNonQuery();
                        logger.Debug("Updated Rezultat: " + entity);
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex)
            {
                logger.Error("Error updating Rezultat: " + entity, ex);
            }
        }
    }
}

File: 2nd Year/2nd Semester/MPP/Triatlon-Java/TriatlonPersistence/src/main/java/triatlon/persistence/RezultatRepository.java
package triatlon.persistence;

import triatlon.model.Arbitru;
import triatlon.model.Participant;
import triatlon.model.Rezultat;
import triatlon.model.TipProba;
import triatlon.persistence.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class RezultatRepository implements IRezultatRepository {

    private static final Logger logger = LogManager.getLogger();
    private final JdbcUtils jdbcUtils;
    private final IParticipantRepository participantRepository;
    private final IArbitruRepository arbitruRepository;

    public RezultatRepository(Properties properties,
                              IParticipantRepository participantRepository,
                              IArbitruRepository arbitruRepository) {
        logger.info("Initializing RezultatRepository with properties: {}", properties);
        jdbcUtils = new JdbcUtils(properties);
        this.participantRepository = participantRepository;
        this.arbitruRepository = arbitruRepository;
    }

    @Override
    public Rezultat findOne(Integer integer) {
        logger.traceEntry("finding Rezultat with id {}", integer);
        String sql = "SELECT * FROM Rezultat WHERE id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, integer);

            try (ResultSet rs = preparedStatement.executeQuery()) {
                if (rs.next()) {
                    int participantId = rs.getInt("participant_id");
                    int arbitruId = rs.getInt("arbitru_id");
                    String tipProbaStr = rs.getString("tip_proba");
                    int punctaj = rs.getInt("punctaj");

                    Participant participant = participantRepository.findOne(participantId);
                    Arbitru arbitru = arbitruRepository.findOne(arbitruId);
                    TipProba tipProba = TipProba.valueOf(tipProbaStr);

                    return new Rezultat(
                            rs.getInt("id"),
                            participant,
                            arbitru,
                            tipProba,
                            punctaj
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        return null;
    }

    @Override
    public Iterable<Rezultat> findAll() {
        List<Rezultat> rezultate = new ArrayList<>();
        logger.traceEntry("finding all Rezultate");
        String sql = "SELECT * FROM Rezultat";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql);
             ResultSet resultSet = preparedStatement.executeQuery()) {

            while (resultSet.next()) {
                int participantId = resultSet.getInt("participant_id");
                int arbitruId = resultSet.getInt("arbitru_id");
                String tipProbaStr = resultSet.getString("tip_proba");
                int punctaj = resultSet.getInt("punctaj");

                Participant participant = participantRepository.findOne(participantId);
                Arbitru arbitru = arbitruRepository.findOne(arbitruId);
                TipProba tipProba = TipProba.valueOf(tipProbaStr);

                rezultate.add(new Rezultat(
                        resultSet.getInt("id"),
                        participant,
                        arbitru,
                        tipProba,
                        punctaj
                ));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
        return rezultate;
    }

    @Override
    public void save(Rezultat entity) {
        logger.traceEntry("saving Rezultat {}", entity);
        String sql = "INSERT INTO Rezultat (participant_id, arbitru_id, tip_proba, punctaj) VALUES (?, ?, ?, ?)";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {

            preparedStatement.setInt(1, entity.getParticipant().getId());
            preparedStatement.setInt(2, entity.getArbitru().getId());
            preparedStatement.setString(3, entity.getTipProba().name());
            preparedStatement.setInt(4, entity.getPunctaj());

            int affectedRows = preparedStatement.executeUpdate();
            if (affectedRows > 0) {
                try (ResultSet generatedKeys = preparedStatement.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        entity.setId(generatedKeys.getInt(1));
                    }
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer integer) {
        logger.traceEntry("deleting Rezultat with id {}", integer);
        String sql = "DELETE FROM Rezultat WHERE id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, integer);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void update(Rezultat entity) {
        logger.traceEntry("updating Rezultat {}", entity);
        String sql = "UPDATE Rezultat SET participant_id = ?, arbitru_id = ?, tip_proba = ?, punctaj = ? WHERE id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, entity.getParticipant().getId());
            preparedStatement.setInt(2, entity.getArbitru().getId());
            preparedStatement.setString(3, entity.getTipProba().name());
            preparedStatement.setInt(4, entity.getPunctaj());
            preparedStatement.setInt(5, entity.getId());

            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }
}